src/metabase/query_processor/middleware/desugar.clj

```diff
@@ -4,21 +4,54 @@
    [metabase.legacy-mbql.predicates :as mbql.preds]
    [metabase.legacy-mbql.schema :as mbql.s]
    [metabase.legacy-mbql.util :as mbql.u]
+   [metabase.lib.metadata :as lib.metadata]
    [metabase.lib.util.match :as lib.util.match]
+   [metabase.query-processor.store :as qp.store]
    [metabase.util.malli :as mu]))
 
+(defn- add-missing-field-base-types
+  "Add base type to fields with id that are missing it. It is necessary for correct function
+   of [[metabase.legacy-mbql.util/desugar-is-empty-and-not-empty]]."
+  [query]
+  (lib.util.match/replace query
+    (field-clause :guard (fn [clause]
+                           (and (mbql.preds/Field? clause)
+                                (integer? (second clause))
+                                (not (contains? (get clause 2) :base-type)))))
+    (let [id (second field-clause)
+          metadata (lib.metadata/field (qp.store/metadata-provider) id)
+          {:keys [base-type]} metadata]
+      (update field-clause 2 assoc
+              :base-type base-type
+              ::desugar-added-base-type true))))
+
+(defn- remove-desugar-added-base-types
+  [query]
+  (lib.util.match/replace query
+    (field-clause :guard (fn [clause]
+                           (and (mbql.preds/Field? clause)
+                                (integer? (second clause))
+                                (get-in clause [2 ::desugar-added-base-type]))))
+    (update field-clause 2 (comp not-empty
+                                 #(dissoc % :base-type ::desugar-added-base-type)))))
```

Comment:
I think this would be easier to read with some destructuring in the `:guard`, `(fn [[_field id opts :as clause]] ...)`.

Reply:
Guard fn could be called on something that's not destructurable. That would result in an exception. Hence second and get-in must be called after the Field? check.